File: block/python/aes_counter.py
#!/usr/bin/python

import argparse
import struct
import sys
import time
from Crypto.Cipher import AES
from Crypto.Protocol import KDF
from Crypto.Hash import RIPEMD, SHA

parser = argparse.ArgumentParser(description='AES in counter mode CSPRNG')
parser.add_argument('-b','--binary', help='Raw binary output.', action="store_true")
parser.add_argument('-k','--key', help='AES key.')
parser.add_argument('-n','--numbers', help='Quantity of random numbers.')
args = parser.parse_args()

def sxor(s1, s2):
    return ''.join(chr(ord(a)^ord(b)) for a,b in zip(s1,s2))

with open('/proc/interrupts','r') as f:
    data = f.read().replace('\n','')

digest = RIPEMD.new(data)
ripemd = digest.hexdigest()

digest = SHA.new(data)
sha = digest.hexdigest()

salt = sxor(sha, ripemd)
key = KDF.PBKDF2(sha, salt, 16, 0)
number = 1

if args.key:
    key = KDF.PBKDF2(bytes(args.key), salt, 16, 0)

if args.numbers:
    number = int(args.numbers)

aes = AES.new(key)
counter = int(time.time())

# AES in counter mode
for i in xrange(0, number):
    out = aes.encrypt(repr(counter).zfill(16))
    counter += 1
    if args.binary:
        sys.stdout.write(out)
    else:
        res = struct.unpack('QQ', out)
        print(res[0]*2**64+res[1])

File: block/javascript/aes_counter.js
#!/usr/bin/node

var BigNumber = require('bignumber');
var crypto = require('crypto');
var stdio = require('stdio');

var opts = stdio.getopt({
    'binary': {key: 'b', args: 0, description: 'Raw binary output.'},
    'key': {key: 'k', args: 1, description: 'AES key.'},
    'numbers': {key: 'n', args: 1, description: 'Quantity of random numbers.'}
});

function encrypt(s, k) {
    var aes = crypto.createCipher('aes128', k);
    return aes.update(s);
}

var salt = 'ce7b67344bcf92a881ea392ed0904fa6'; // for pbkdf2
var key = crypto.pbkdf2Sync('59947d9ef5fc733f95ea300c1ec8fa11', salt, 0, 16);
var number = 1;

if(opts.key) key = crypto.pbkdf2Sync(opts.key, salt, 0, 16);
if(opts.numbers) number = opts.numbers;

// Almost never return exponential notation
BigNumber.config({ EXPONENTIAL_AT: 1e+9 });

// Start nanoseconds benchmarking
var startTime = process.hrtime();
var counter = Date.now();

for (var i=0; i<number; i++) {
    var out = new Buffer(encrypt(('0000000000000000'+counter).slice(-16), key));
    counter++;
    if(opts.binary) process.stdout.write(out);
    else console.log(new BigNumber(out.toString('hex'), 16).toString());
}
